### PR TITLE
Adding Boskos / Mason configuration and tools

### PR DESCRIPTION
**Pull request by sebastienvas**
In order to review, please look at the commits as the vendor update hides all the changes.

### CODE REVIEW

**Comment by ldemailly** [comment]
screen shot 2018-02-14 at 11 50 16 pm
👍 

too bad the repo history will carry those forever ;-)

File: boskos/cmd/mason_client/main.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+		return
+	}
+	client := masonClient{mason: mason.NewClient(client.NewClient(*owner, *client.BoskosURL))}
+	if *kubecfgPath == "" {
+		logrus.Panic("flag --type must be set")
+	}
+	c1, acquireCancel := context.WithTimeout(context.Background(), timeout)
+	c2, updateCancel := context.WithCancel(context.Background())
+	defer acquireCancel()
+	defer updateCancel()
```
Comment (chxchx) [inline boskos/cmd/mason_client/main.go]: I think stacked defers are executed in LIFO order so I think it makes more sense to move `defer updateCancel()` after error checking on acquire().

File: boskos/gcp/gce.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+			{
+				AccessConfigs: []*compute.AccessConfig{
+					{
+						Name: "External NAT",
+						Type: oneToOneNAT,
+					},
+				},
+			},
+		},
+		ServiceAccounts: []*compute.ServiceAccount{
```
Comment (chxchx) [inline boskos/gcp/gce.go:97]: I thought you need something like
```
                ServiceAccounts: []*compute.ServiceAccount{
			&compute.ServiceAccount{
				Email:  "default",
				Scopes: config.Scopes,
			},
		},
```
so we know the element we put in the slice is a pointer?

File: boskos/gcp/gcloud.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+func runCommand(name string, args ...string) error {
```
Comment (chxchx) [inline boskos/gcp/gcloud.go]: Why not use the existing Shell() utility we have in test-infra?
If you insist, perhaps make this a util as well?

File: boskos/gcp/gke.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(defaultSleepTime):
+			newOp, err := cc.service.Projects.Zones.Operations.Get(project, zone, op.Name).Context(ctx).Do()
+			if err != nil {
+				return err
+			}
+			switch newOp.Status {
+			case "DONE":
```
Comment (chxchx) [inline boskos/gcp/gke.go]: Are these string standardized? Maybe define they as enum instead of hardcoding?

File: boskos/gcp/test-configs.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+          zone: us-central-1f
+        vms:
+        - machinetype: n1-standard-4
+          sourceimage: projects/debian-cloud/global/images/debian-9-stretch-v20180105
+          zone: us-central-1f
+          tags:
+          - http-server
+          - https-server
+          scopes:
+          - https://www.googleapis.com/auth/cloud-platform
```
Comment (chxchx) [inline boskos/gcp/test-configs.yaml]: pro tip: config your editor to always add new line at the end

File: scripts/set-boskos-project.sh [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+. ${DIR}/all-utilities || { echo "Cannot load Bash utilities" ; exit 1 ; }
+
+# Exit immediately for non zero status
+set -e
+# Check unset variables
+set -u
+# Print commands
+set -x
+
+USERS=()
```
Comment (chxchx) [inline scripts/set-boskos-project.sh]: just curious, how is this different from
```
USERS=(
    'serviceAccount:boskos@istio-testing.iam.gserviceaccount.com'
    'serviceAccount:istio-prow-test-job@istio-testing.iam.gserviceaccount.com'
    'group:mdb.istio-testing@google.com'
)
```

File: scripts/update-e2e-cluster.sh [hunk 7]
```diff
@@ -59,43 +55,45 @@ if [ "${REPO}" != 'istio' ] && [ "${REPO}" != 'daily-release' ]; then
...
 
 # Try to create a rotation cluster, named $REPO-e2e-rbac-rotation-<suffix>, suffix can be 1 or 2
 gcloud config set container/use_client_certificate True
-for i in {1..2}
-do
+for i in {1..2}; do
   CLUSTER_NAME="${REPO}-e2e-rbac-rotation-${i}"
-  result=$(gcloud container clusters create ${CLUSTER_NAME} --zone ${ZONE} --project ${PROJECT_NAME} --cluster-version ${CLUSTER_VERSION} \
-  --machine-type ${MACHINE_TYPE} --num-nodes ${NUM_NODES} --no-enable-legacy-authorization --enable-kubernetes-alpha --quiet \
+  // Passing KUBECONFIG as an env will override default ~/.kube/config
```
Comment (chxchx) [inline scripts/update-e2e-cluster.sh]: Is this meant to be a comment? Start with `#` instead?

**Review by chxchx — COMMENTED** [review]
Now I see why checking in vendor might be a bad idea

File: boskos/gcp/gce.go (hunk 2 above)
Comment (sebastienvas) [reply boskos/gcp/gce.go:97]: gometalinter will complain if you do that.

File: boskos/cmd/mason_client/main.go (hunk 1 above)
Comment (sebastienvas) [reply boskos/cmd/mason_client/main.go]: Done

File: boskos/gcp/test-configs.yaml (hunk 5 above)
Comment (sebastienvas) [reply boskos/gcp/test-configs.yaml]: Done

File: boskos/cmd/mason/main.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2017 Istio Authors
```
Comment (rkpagadala) [inline boskos/cmd/mason/main.go]: change 2017 to 2018 here and elsewhere for new files.

File: boskos/cmd/mason_client/main.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
+// Copyright 2017 Istio Authors
```
Comment (rkpagadala) [inline boskos/cmd/mason_client/main.go]: 2018

File: boskos/gcp/gke.go (hunk 4 above)
Comment (rkpagadala) [reply boskos/gcp/gke.go]: are there no other values? default?

File: boskos/README.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+resources and transition between different states.
+
+In istio, we are using Boskos to pre-provision cluster and vms on GCP. This
+functionality was added with the Mason module.
+
+## Setup
+
+Boskos runs on a kubernetes cluster. Initially only Prow jobs will be using
+Boskos. Since Boskos service is only available intra cluster, it is important
+that Boskos runs in the same cluster were test jobs are running.
```
Comment (rkpagadala) [inline boskos/README.md]: where test jobs are running.

File: boskos/README.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+```bash
+make mason-image
+# Update mason deployment to point to the new image, create a PR
+make mason-deployment
+```
+
+## Mason
+
+Mason is used for resources that have resources dependency. Mason resource are
+virtual, they only exist inside other resources. An example is a mason resource
```
Comment (rkpagadala) [inline boskos/README.md]: An example of a mason resource

File: boskos/gcp/gke.go [hunk 12]
```diff
@@ -61,9 +64,9 @@ func (cc *containerEngine) waitForOperation(ctx context.Context, op *container.O
 				return err
 			}
 			switch newOp.Status {
-			case "DONE":
+			case operationDone:
 				return nil
-			case "ABORTING":
+			case operationAborting:
```
Comment (rkpagadala) [inline boskos/gcp/gke.go]: any other operations status that we need to care about? do we need default?

File: boskos/gcp/gcloud.go (hunk 3 above)
Comment (sebastienvas) [reply boskos/gcp/gcloud.go]: Done

File: boskos/cmd/mason/main.go (hunk 8 above)
Comment (sebastienvas) [reply boskos/cmd/mason/main.go]: done

File: boskos/cmd/mason_client/main.go (hunk 9 above)
Comment (sebastienvas) [reply boskos/cmd/mason_client/main.go]: done

File: boskos/gcp/gcloud.go (hunk 3 above)
Comment (sebastienvas) [reply boskos/gcp/gcloud.go]: done

File: boskos/gcp/gke.go (hunk 4 above)
Comment (sebastienvas) [reply boskos/gcp/gke.go]: added logging

**Review by sebastienvas — COMMENTED** [review]
PTAL

File: boskos/README.md (hunk 10 above)
Comment (sebastienvas) [reply boskos/README.md]: done

File: boskos/README.md (hunk 11 above)
Comment (sebastienvas) [reply boskos/README.md]: done

**Comment by rkpagadala** [comment]
/lgtm
assuming you will fix the failure in the tests

**Pull request merged**